src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -3217,37 +3217,120 @@ private void CompileTrappableExpression(List<Expression> exprList, StatementAst
+            return MustSetSuccessAfterEvaluating(expressionAst);
+        }
+
+        /// <summary>
+        /// Determines whether an assignment statement must have an explicit setting
+        /// for $? = $true after it by the compiler.
+        /// </summary>
+        /// <param name="assignmentStatementAst">The assignment statement to examine.</param>
+        /// <returns>True is the compiler should add the success setting, false otherwise.</returns>
+        private bool MustSetSuccessAfterEvaluating(AssignmentStatementAst assignmentStatementAst)
+        {
+            // Get right-most RHS in cases like $x = $y = <expr>
+            StatementAst innerRhsStatementAst = assignmentStatementAst.Right;
+            while (innerRhsStatementAst is AssignmentStatementAst rhsAssignmentAst)
+            {
+                innerRhsStatementAst = rhsAssignmentAst.Right;
+            }
+
+            // Simple assignments to pure expressions may need $? set, so examine the RHS statement for pure expressions
+            switch (innerRhsStatementAst)
+            {
+                case CommandExpressionAst commandExpression:
+                    return MustSetSuccessAfterEvaluating(commandExpression.Expression);
 
-                    pipeAst = right as PipelineAst;
-                    if (right is CommandExpressionAst ||
-                        (pipeAst != null && pipeAst.PipelineElements.Count == 1 &&
-                         pipeAst.PipelineElements[0] is CommandExpressionAst))
+                case PipelineAst rhsPipelineAst:
+                    return MustSetSuccessAfterEvaluating(rhsPipelineAst);
+
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Determines whether an expression in a statement must have an explicit setting
+        /// for $? = $true after it by the compiler.
+        /// </summary>
+        /// <param name="expressionAst">The expression to examine.</param>
+        /// <returns>True is the compiler should add the success setting, false otherwise.</returns>
+        private bool MustSetSuccessAfterEvaluating(ExpressionAst expressionAst)
+        {
+            switch (expressionAst)
+            {
+                case ParenExpressionAst parenExpression:
+                    // Pipelines in paren expressions that are just pure expressions will need $? set
+                    // e.g. ("Hi"), vs (Test-Path ./here.txt)
+                    return MustSetSuccessAfterEvaluating(parenExpression.Pipeline);
+
+                case SubExpressionAst subExpressionAst:
+                    // Subexpressions generally set $? since they encapsulate a statement block
+                    // But $() requires an explicit setting
+                    return subExpressionAst.SubExpression.Statements.Count == 0;
+
+                case ArrayExpressionAst arrayExpressionAst:
+                    switch (arrayExpressionAst.SubExpression.Statements.Count)
```

Comment:
`ArrayExpression` also encapsulate a statement block, so why is it treated differently than `subExpression`?

Reply:
Ah, I'm glad you asked because it probably means I need to add more comments.

We optimise ArrayExpressionAsts differently elsewhere when we don't use the pipeline. So with SubExpressionAsts, if there are any values in it, we set `$?` correctly without added logic. But with an ArrayExpressionAst, if there is only a single expression inside it, we need the extra `$?` logic.

So basically the inconsistency here mirrors the inconsistency in pure expression handling for these expressions elsewhere.

Reply:
I see. Thanks for the clarification. Yes, please add comments about this.

Reply:
Note that `ArrayExpressionAst` is optimized only if `Traps == null`. I don't see the need to check for Traps here, but want to bring it up just in case I'm missing anything.

Reply:
Yeah, I actually played with that too. I tested shortcutting when traps are set (I have tests for it), but this code reflects what I found created the outcome we expect.

Reply:
I just tested this again, above the switch I added:

```csharp
if (arrayExpressionAst.SubExpression.Traps != null)
{
    return false;
}
```

That causes this test to fail:

```powershell
    [-] Sets $? for case 'Write-Error "Bad"; @(trap { continue }); Hook $?' 101ms
      Expected $true, but got $false.
      73:         $script:hookValues | Should -Be $Results
      at <ScriptBlock>, C:\Users\roholt\Documents\Dev\PowerShell\test\powershell\Language\Scripting\DollarHook.Tests.ps1: line 73
```

Reply:
Yes, I expect this test to fail in that case because the statement list is empty. It's the same as the SubExpression case `$()` you dealt with above.
But I think `@(trap { continue } "hello")` or `@(trap { continue } "hello", "world")` will work, since it would go through the same code as the SubExpression in this case.

Reply:
Ah ok. Have added a check for that.